server/quota/quota_manager.go

```diff
@@ -0,0 +1,291 @@
+	// TODO: set up a dedicated redis client for quota
+	store, err := goredisstore.NewCtx(env.GetDefaultRedisClient(), prefix)
+	if err != nil {
+		return nil, status.InternalErrorf("unable to init redis store: %s", err)
+	}
+
+	period := time.Duration(config.PeriodDurationUsec) * time.Microsecond
+	quota := throttled.RateQuota{
+		MaxRate:  throttled.PerDuration(int(config.NumRequests), period),
+		MaxBurst: int(config.MaxBurst),
+	}
+
+	rateLimiter, err := throttled.NewGCRARateLimiterCtx(store, quota)
+	rateLimiter.SetMaxCASAttemptsLimit(maxRedisRetry)
+	if err != nil {
+		return nil, status.InternalErrorf("unable to create GCRARateLimiter: %s", err)
+	}
+
+	bucket := &gcraBucket{
+		config:      config,
+		rateLimiter: rateLimiter,
+	}
+
+	return bucket, nil
+}
+
+type namespace struct {
+	name   string
+	config *namespaceConfig
+
+	defaultBucket Bucket
+	bucketsByKey  map[string]Bucket
+}
+
+type bucketCreatorFn func(environment.Env, *tables.QuotaBucket) (Bucket, error)
+
+type QuotaManager struct {
+	env           environment.Env
+	namespaces    map[string]*namespace
+	bucketCreator bucketCreatorFn
+}
+
+func NewQuotaManager(env environment.Env) (*QuotaManager, error) {
+	return newQuotaManager(env, createGCRABucket)
+}
+
+func newQuotaManager(env environment.Env, bucketCreator bucketCreatorFn) (*QuotaManager, error) {
+	config, err := fetchConfigFromDB(env)
+	if err != nil {
+		return nil, err
+	}
+	qm := &QuotaManager{
+		env:           env,
+		namespaces:    make(map[string]*namespace),
+		bucketCreator: bucketCreator,
+	}
+
+	for nsName, nsConfig := range config {
+		ns, err := qm.createNamespace(env, nsName, nsConfig)
+		if err != nil {
+			return nil, err
+		}
+		qm.namespaces[nsName] = ns
+	}
+
+	return qm, nil
+}
+
+func (qm *QuotaManager) createNamespace(env environment.Env, name string, config *namespaceConfig) (*namespace, error) {
+	defaultBucketConfig := config.bucketsByName[defaultBucketName]
+	if defaultBucketConfig == nil {
+		return nil, status.InvalidArgumentErrorf("default quota bucket is unset in namespace: %q", name)
+	}
+
+	defaultBucket, err := qm.bucketCreator(env, defaultBucketConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := &namespace{
+		name:          name,
+		config:        config,
+		defaultBucket: defaultBucket,
+		bucketsByKey:  make(map[string]Bucket),
+	}
+
+	for _, bc := range config.bucketsByName {
+		bucket, err := qm.bucketCreator(env, bc)
+		if err != nil {
+			return nil, err
+		}
+
+		keys := config.quotaKeysByBucketName[bc.Name]
+		for _, key := range keys {
+			ns.bucketsByKey[key] = bucket
+		}
+	}
+	return ns, nil
+}
+
+// findBucket finds the bucket given a namespace and key. If the key is found in
+// bucketsByKey map, return the corresponding bucket. Otherwise, return the
+// default bucket. Returns nil if the namespace is not found.
+func (qm *QuotaManager) findBucket(namespace string, key string) Bucket {
+	ns, ok := qm.namespaces[namespace]
```

Comment:
i am surprised you don't need any locking around namespaces or bucketsByKey -- have you run this through the race detector?

Reply:
blaze test --@io_bazel_rules_go//go/config:race ...

Reply:
Currently these maps were only written on start up. Once I implement the pubsub stuff, we can add locks.